README.md
# synupkeep [![builds.sr.ht status][build-status-badge]][build-status]

> A program to upkeep Synapse servers

This is a simple script to upkeep [Synapse][synapse] servers. It deletes
events, local media and cached remote media up to a specified point in time.
Old user and room avatars are (for now) excluded from getting purged due to
issues with the latest ones getting removed as well. Currently, only Synapse
instances using a [PostgreSQL][postgresql] database are supported.

Note that content on remote (federated) servers can obviously not be deleted
and that Synapse might not consider the most recent content as _history_, which
excludes it from being deleted even if it is technically inside the specified
purge timeframe.

## Table of contents

+ [Install](#install)
  + [Installing with Docker](#installing-with-docker)
  + [Installing without Docker](#installing-without-docker)
  + [Dependencies](#dependencies)
  + [Updating](#updating)
    + [Updating with Docker](#updating-with-docker)
    + [Updating without Docker](#updating-without-docker)
+ [Usage](#usage)
  + [Running with Docker](#running-with-docker)
  + [Running without Docker](#running-without-docker)
  + [Configuration](#configuration)
+ [Maintainer](#maintainer)
+ [Contribute](#contribute)
+ [License](#license)

## Install

The recommended way to run is via [Docker][docker]. Basic instructions on how
to run without it are also provided.

### Installing with Docker

To install for running with Docker, you can simply pull the prebuilt image from
[Docker Hub][docker-hub]:

```zsh
user@local:~$ docker pull mserajnik/synupkeep
```

Alternatively, you can also build the image yourself. The user that is used
inside the container has UID `1000` and GID `1000` by default. You can adjust
this (e.g., to match your host UID/GID) by providing the arguments `USER_ID`
and `GROUP_ID` when making a build.

### Installing without Docker

To install without Docker, you can simply clone the repository and install
dependencies using Poetry.

```zsh
user@local:~$ git clone https://git.sr.ht/~mser/synupkeep
user@local:~$ cd synupkeep
user@local:synupkeep$ poetry install
```

### Dependencies

+ [Docker][docker] (when running with Docker)
+ [Python 3.7+][python] (when running without Docker)
+ [Poetry][poetry] (when running without Docker)

### Updating

This script follows [semantic versioning][semantic-versioning] and any breaking
changes that require additional attention will be released under a new major
version (e.g., `2.0.0`). Minor version updates (e.g., `1.1.0` or `1.2.0`) are
therefore always safe to simply install.

When necessary, this section will be expanded with upgrade guides for new major
versions.

#### Updating with Docker

Simply pull the latest Docker image to update:

```zsh
user@local:~$ docker pull mserajnik/synupkeep
```

#### Updating without Docker

If you have installed via cloning the repository, you can update using Git as
well:

```zsh
user@local:synupkeep$ git pull
```

## Usage

### Running with Docker

To make running with Docker as easy as possible, a working
[Docker Compose][docker-compose] example setup is provided. To get started with
this example setup, simply duplicate `docker-compose.yml.example` as
`docker-compose.yml` and adjust the path to the Synapse data directory in the
`volumes` section, the variables in the `environment` section as described
[here](#configuration) and change the `command` if you want to (you can either
run the script once or periodically via cron job).

Take note of the time zone set via the `TZ` environment variable. This is
particularly important for the cron job to run at the time you expect it to.
The time zone has to be set in the
[tz database format][tz-database-time-zones].

Finally, start the container:

```zsh
user@local:synupkeep$ docker-compose up -d
```

Depending on your choice, the script will now either run once (with command
`run`) or periodically via cron job until stopped (with command `cron`).

The user that is used inside the container when the script is run has the UID
`1000` and the GID `1000` by default. You can change these by providing the
environment variables `CUSTOM_UID` and `CUSTOM_GID` when creating a container.

### Running without Docker

The easiest way to run synupkeep without Docker is via Poetry:

```zsh
user@local:synupkeep$ poetry run synupkeep
```

You can use either environment variables or arguments to configure the script
when running without Docker. The script will look for the provided argument
first and fall back to the respective environment variable if the argument is
not provided.

To see the required and optional arguments, simple launch the script with the
`-h` flag:

```zsh
user@local:synupkeep$ poetry run synupkeep -h
usage: synupkeep [-h] [--delta DELTA] [--logging-level {CRITICAL,WARNING,INFO,TRACE,SUCCESS,ERROR,DEBUG}] [--api-url API_URL] [--media-store MEDIA_STORE] [postgres_connection_string] [synapse_auth_token]

positional arguments:
  postgres_connection_string
  synapse_auth_token

optional arguments:
  -h, --help            show this help message and exit
  --delta DELTA, -d DELTA
  --logging-level {CRITICAL,WARNING,INFO,TRACE,SUCCESS,ERROR,DEBUG}, -l {CRITICAL,WARNING,INFO,TRACE,SUCCESS,ERROR,DEBUG}
  --api-url API_URL, -a API_URL
  --media-store MEDIA_STORE, -m MEDIA_STORE
```

### Configuration

Configuration is done via environment variables when running with Docker. These
can also be used when running without Docker, although using launch parameters
instead might be preferable in that case (see [here](#running-without-docker)).

+ `SYNUPKEEP_MEDIA_STORE_PATH=/data/media_store`: the path to the Synapse
  `media_store` directory.
+ `SYNUPKEEP_POSTGRES_CONNECTION_STRING=`: the PostgreSQL connection string for
  the Synapse database, e.g.,
  `host=localhost port=5432 user=synapse dbname=synapse password=password`.
+ `SYNUPKEEP_SYNAPSE_API_URL=http://localhost:8008/`: the base URL of the
  Synapse API. __Trailing `/` required.__
+ `SYNUPKEEP_SYNAPSE_AUTH_TOKEN=`: the access token of an admin user. Can be
  created via
  `curl -XPOST -d '{"type":"m.login.password", "user":"<userid>", "password":"<password>"}' "http://localhost:8008/_matrix/client/r0/login"`.
+ `SYNUPKEEP_DELTA=86400`: the current date minus the provided value here (in
  seconds) is the date up to which the script will purge.
+ `SYNUPKEEP_LOGGING_LEVEL=INFO`: the desired logging level. Has to be one on
  these: `TRACE`, `DEBUG`, `INFO`, `SUCCESS`, `WARNING`, `ERROR`, `CRITICAL`.
+ `SYNUPKEEP_DOCKER_CRON_SCHEDULE=0 4 * * *`: the cron schedule for running
  with Docker using the `cron` command.

## Maintainer

[Michael Serajnik][maintainer]

## Contribute

You are welcome to help out!

[Open a ticket][tickets] or [send a patch][patches].

## License

[AGPLv3](LICENSE) © Michael Serajnik

[synapse]: https://github.com/matrix-org/synapse/
[postgresql]: https://www.postgresql.org/
[docker]: https://www.docker.com/
[docker-hub]: https://hub.docker.com/r/mserajnik/synupkeep/
[python]: https://www.python.org/
[poetry]: https://python-poetry.org/
[semantic-versioning]: https://semver.org/
[docker-compose]: https://docs.docker.com/compose/
[tz-database-time-zones]: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

[build-status]: https://builds.sr.ht/~mser/synupkeep
[build-status-badge]: https://builds.sr.ht/~mser/synupkeep.svg

[maintainer]: https://sr.ht/~mser/
[tickets]: https://todo.sr.ht/~mser/synupkeep
[patches]: https://lists.sr.ht/~mser/public-inbox


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2021-04-03

### Added

+ Initial release

[Unreleased]: https://git.sr.ht/~mser/synupkeep/tree/develop
[1.0.0]: https://git.sr.ht/~mser/synupkeep/tree/1.0.0


synupkeep/__init__.py
# synupkeep
# Copyright (C) 2021-present  Chris Braun  https://github.com/cryzed
# Copyright (C) 2021-present  Michael Serajnik  https://sr.ht/~mser/

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

__version__ = '1.0.0'


import argparse
import enum
import json
import os
import sys
import time
import urllib.parse
from datetime import datetime, timedelta
from typing import Optional, Set, List, Dict

import postgres
import psycopg2
import requests
from loguru import logger

PURGE_HISTORY_API_ENDPOINT = "/_synapse/admin/v1/purge_history/{room}"
PURGE_HISTORY_STATUS_API_ENDPOINT = "/_synapse/admin/v1/purge_history_status/{id}"
PURGE_REMOTE_MEDIA_API_ENDPOINT = "/_synapse/admin/v1/purge_media_cache"


class ExitCode(enum.IntEnum):
    Success = 0
    Failure = 1


class PurgeStatus(str, enum.Enum):
    Active = "active"
    Complete = "complete"
    Failed = "failed"


def get_argument_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "postgres_connection_string",
        nargs="?",
        default=os.getenv("SYNUPKEEP_POSTGRES_CONNECTION_STRING"),
    )
    parser.add_argument(
        "synapse_auth_token",
        nargs="?",
        default=os.getenv("SYNUPKEEP_SYNAPSE_AUTH_TOKEN"),
    )
    parser.add_argument(
        "--delta",
        "-d",
        type=int,
        default=os.getenv("SYNUPKEEP_DELTA", 60 * 60 * 24),
    )
    parser.add_argument(
        "--logging-level",
        "-l",
        choices={"TRACE", "DEBUG", "INFO", "SUCCESS", "WARNING", "ERROR", "CRITICAL"},
        default=os.getenv("SYNUPKEEP_LOGGING_LEVEL", "INFO"),
    )
    parser.add_argument(
        "--api-url",
        "-a",
        default=os.getenv("SYNUPKEEP_SYNAPSE_API_URL", "http://localhost:8008/"),
    )
    parser.add_argument(
        "--media-store",
        "-m",
        default=os.getenv("SYNUPKEEP_MEDIA_STORE_PATH", "/data/media_store"),
    )
    return parser


def get_api_url(base_url: str, suffix: str) -> str:
    return urllib.parse.urljoin(base_url, suffix)


def get_delta_date(seconds: int) -> datetime:
    return datetime.utcnow() - timedelta(seconds=seconds)


def get_last_event_id(
    db: postgres.Postgres, room_id: str, before: datetime
) -> Optional[str]:
    return db.one(
        "SELECT event_id FROM events WHERE received_ts <= %(before)s AND room_id=%(room_id)s ORDER BY received_ts DESC LIMIT 1",
        before=int(before.timestamp() * 1000),
        room_id=room_id,
    )


def get_important_media_ids(db: postgres.Postgres) -> Set[str]:
    media_ids = set()

    for event in db.all("SELECT json FROM event_json"):
        data = json.loads(event)
        content = data["content"]
        type_ = data["type"]

        if type_ == "m.room.member":
            avatar_url = content.get("avatar_url")
            if avatar_url:
                media_ids.add(avatar_url)
        elif type_ == "m.room.avatar":
            if "url" not in content:
                logger.warning("No URL defined for {!r}-event: {!r}", type_, data)
                continue

            media_ids.add(content["url"])

    return set(urllib.parse.urlsplit(url).path[1:] for url in media_ids)


def get_local_media_record_ids(db: postgres.Postgres, before: datetime) -> List[str]:
    return db.all(
        "SELECT media_id FROM local_media_repository WHERE created_ts <= %(before)s",
        before=int(before.timestamp() * 1000),
    )


def get_room_record_ids(db: postgres.Postgres) -> List[str]:
    return db.all("SELECT room_id FROM rooms")


def delete_local_media_record(db: postgres.Postgres, media_id: str) -> None:
    db.run(
        "DELETE FROM local_media_repository WHERE media_id = %(media_id)s",
        media_id=media_id,
    )


def purge_history(
    session: requests.Session, api_url: str, room_id: str, event_id: str
) -> Optional[str]:
    url = get_api_url(api_url, PURGE_HISTORY_API_ENDPOINT).format(room=room_id)
    payload = {"delete_local_events": True, "purge_up_to_event_id": event_id}
    return session.post(url, json=payload).json().get("purge_id")


def purge_history_status(
    session: requests.Session, api_url: str, purge_id: str
) -> PurgeStatus:
    url = get_api_url(api_url, PURGE_HISTORY_STATUS_API_ENDPOINT).format(id=purge_id)
    return PurgeStatus(session.get(url).json()["status"])


def purge_remote_media(
    session: requests.Session, api_url: str, before: datetime
) -> Dict:
    url = get_api_url(api_url, PURGE_REMOTE_MEDIA_API_ENDPOINT)
    return session.post(
        url, params={"before_ts": int(before.timestamp() * 1000)}
    ).json()


def get_local_media_paths(media_store_path: str, media_id: str) -> List[str]:
    paths = []
    for folder_name in "local_content", "local_thumbnails":
        path = os.path.join(
            media_store_path, folder_name, media_id[:2], media_id[2:4], media_id[4:]
        )
        paths.append(path)

    return paths


def wait_for_purge(
    session: requests.Session, api_url: str, purge_id: str, interval: float = 1
) -> PurgeStatus:
    while True:
        status = purge_history_status(session, api_url, purge_id)
        if status != PurgeStatus.Active:
            return status

        time.sleep(interval)


def run(arguments: argparse.Namespace) -> ExitCode:
    logger.remove()
    logger.add(sys.stderr, level=arguments.logging_level)
    logger.debug("Called with {!r}", arguments)

    if not (arguments.postgres_connection_string and arguments.synapse_auth_token):
        logger.error(
            "Postgres connection string and Synapse auth token can not be empty"
        )
        return ExitCode.Failure

    try:
        db = postgres.Postgres(arguments.postgres_connection_string)
    except (psycopg2.OperationalError, psycopg2.InternalError):
        logger.exception(
            "Connecting to database using {!r} failed:",
            arguments.postgres_connection_string,
        )
        return ExitCode.Failure

    session = requests.session()
    session.headers["Authorization"] = "Bearer " + arguments.synapse_auth_token

    before_date = get_delta_date(int(arguments.delta))
    before_date_string = before_date.isoformat(sep=" ", timespec="seconds")
    logger.info("Purging events up to: {} UTC", before_date_string)

    # Purge room history for all rooms
    rooms = get_room_record_ids(db)
    room_count = len(rooms)
    for index, room_id in enumerate(rooms, start=1):
        logger.info("({}/{}) Processing room: {!r}...", index, room_count, room_id)
        event_id = get_last_event_id(db, room_id, before_date)
        if event_id is None:
            logger.info(
                "No event ID before: {} UTC for room: {!r}, skipping",
                before_date_string,
                room_id,
            )
            continue

        logger.info(
            "Last event ID before: {} UTC for room {!r}: {!r}",
            before_date_string,
            room_id,
            event_id,
        )
        purge_id = purge_history(session, arguments.api_url, room_id, event_id)
        if not purge_id:
            logger.warning("Failed to purge room: {!r}: received no purge ID", room_id)
            continue

        logger.info("Purging room: {!r} in progress: {!r}...", room_id, purge_id)
        result = wait_for_purge(session, arguments.api_url, purge_id)
        logger.info("Purged room: {!r} with status: {!r}", room_id, result)

    logger.info("Purging local media older than: {} UTC...", before_date_string)
    local_media = get_local_media_record_ids(db, before_date)
    important_files = get_important_media_ids(db)

    # Purge local media manually
    old_media = set(local_media) - important_files
    old_media_count = len(old_media)
    logger.info("{} media to be cleaned...", len(old_media))
    for index, media_id in enumerate(old_media, start=1):
        logger.info(
            "({}/{}) Processing media: {!r}...", index, old_media_count, media_id
        )
        paths = get_local_media_paths(arguments.media_store, media_id)
        for path in paths:
            if not os.path.isfile(path):
                logger.debug("{!r} could not be found or is not a file", path)
                continue
            os.remove(path)

        delete_local_media_record(db, media_id)

    # Purge remote media
    logger.info("Purging cached remote media older than: {} UTC...", before_date_string)
    result = purge_remote_media(session, arguments.api_url, before_date)
    logger.info("Purged cached remote media: {!r}", result)
    return ExitCode.Success


def main():
    parser = get_argument_parser()
    arguments = parser.parse_args()
    parser.exit(run(arguments))
